distributed/worker_state_machine.py

```diff
@@ -353,8 +354,60 @@ class AddKeysMsg(SendMessageToScheduler):
 
 @dataclass
 class StateMachineEvent:
-    __slots__ = ("stimulus_id",)
+    __slots__ = ("stimulus_id", "handled")
     stimulus_id: str
+    #: timestamp of when the event was handled by the worker
+    # TODO switch to @dataclass(slots=True) and uncomment (requires Python >=3.10)
+    # handled: float | None = field(init=False, default=None)
+    _classes: ClassVar[dict[str, type[StateMachineEvent]]] = {}
+
+    def __new__(cls, *args, **kwargs):
+        self = object.__new__(cls)
+        self.handled = None
+        return self
+
+    def __init_subclass__(cls):
+        StateMachineEvent._classes[cls.__name__] = cls
+
+    def log(self, *, handled: float) -> StateMachineEvent:
+        """Produce a variant version of self that is small enough to be stored in memory
+        in the medium term and contains meaningful information for debugging
+        """
+        self.handled = handled
+        return self
+
+    def _to_dict(self, *, exclude: Container[str] = ()) -> dict:
```

Comment:
This dictionary conversion seems necessary because `stimulus_log: StateMachineEvent` has been added to `Worker` and thus must be supported by `Worker._to_dict`